---
title: "Plotting DESeq2 Data"
author: "Virginia Belvedere"
date: "September 2025"
format: html
editor: visual
output: html_document
self-contained: true
toc: true
toc-expand: 2
toc_float: true
toc-title: Contents
toc-location: left
number-sections: true
number-depth: 3
---
```

## Plots

```{r}
suppressPackageStartupMessages({
     library(DESeq2)
     library(dplyr)
     library(stringr)
     library(ggplot2)
     library(ggrepel)
     library(pheatmap)
     library(patchwork)
})

outdir <- file.path("RNAseq_project", "BR3", "plots")
if (!dir.exists(outdir)) dir.create(outdir, recursive = TRUE)
```

The following code is for generating a pack of plots which may be useful for downstream work/presenting, including:

-   MA plot

-   p-value and padj histograms

-   sample correlation and top gene heatmaps

-   volcano (with + without target gene highlight)

-   top up/down regulation tables

-   genome scatter of significant DEGs

-   top 20 up/down barplots

```{r}
## tidy DESeq2 results with MSMEG order - this was useful for me as my data would not organise sensibly

suppressPackageStartupMessages({
  library(dplyr)
  library(stringr)
})

as_tidy_res <- function(res) {
  df <- as.data.frame(res)
  df$gene_raw <- rownames(df)

  df %>%
    mutate(
      gene_id = str_extract(gene_raw, "MSMEG_\\d+"),
      genome_order = suppressWarnings(as.numeric(str_remove(gene_id, "MSMEG_"))),
      neg_log10_padj = ifelse(!is.na(padj) & padj > 0, -log10(padj), NA_real_),
      sig = !is.na(padj) & padj < 0.05
    )
}

```

### MA plot

```{r}

## MA plot (normalised counts v. log10pvalue)
# alpha=0.05` colors  genes red by default.

DESeq2::plotMA(res,
               main = "sgRNApos_ATc vs sgRNAneg_ATc (BR1+BR2+BR3)",
               ylim = c(-3, 3),
               alpha = 0.05)

```

### P-value and padj histograms

```{r}
# p-value and FDR histograms
# A big spike near 1 in `padj` is common when few genes are truly DE and after multiple-testing correction.

hist(res$pvalue, breaks = 50, col = "grey",
     main = "p-value distribution (sgRNApos_ATc vs sgRNAneg_ATc)",
     xlab = "p-value")

hist(res$padj, breaks = 50, col = "grey",
     main = "Adjusted p-value (FDR) distribution",
     xlab = "Adjusted p-value")

```

### Volcano plot

```{r}
suppressPackageStartupMessages({
  library(ggplot2)
})

volcano_pvalue_highlight_fdr <- function(res, title,
                                        p_cut = 0.05,
                                        fdr_cut = 0.05,
                                        ycap = 10) {

  df <- as.data.frame(res)
  df$gene <- rownames(df)

  df <- df[is.finite(df$log2FoldChange) & is.finite(df$pvalue), ]

  df$neglog10p <- -log10(df$pvalue)
  df$fdr_sig   <- !is.na(df$padj) & df$padj < fdr_cut

  ggplot(df, aes(x = log2FoldChange, y = pmin(neglog10p, ycap))) +
    # all genes
    geom_point(color = "grey70", size = 1.2, alpha = 0.8) +
    # overlay FDR significant genes
    geom_point(data = df[df$fdr_sig, ],
               color = "red", size = 1.6, alpha = 0.9) +
    # p-value cutoff line
    geom_hline(yintercept = -log10(p_cut), linetype = "dashed", color = "green3") +
    labs(title = title,
         x = expression(log[2]~fold~change),
         y = expression(-log[10]~pvalue)) +
    theme_classic(base_size = 12)
}

p <- volcano_pvalue_highlight_fdr(res, "sgRNApos_ATc vs sgRNAneg_ATc")
print(p)

```

### Top up/down regulated genes

```{r}
get_top_degs <- function(res, n = 20) {
     df <- as.data.frame(res) %>%
         mutate(gene = rownames(.)) %>%
         filter(!is.na(padj)) %>%
         arrange(padj)
     
     top_up   <- df %>% filter(log2FoldChange > 0) %>% head(n)
     top_down <- df %>% filter(log2FoldChange < 0) %>% head(n)
     
     list(up = top_up, down = top_down)
}
top_degs <- get_top_degs(res, n = 20)
top_degs$up
top_degs$down

```

### Whole genome scatter graph (bit random but I like it as a visual)

```{r}
tidy_res <- as_tidy_res(res)   # use UNshrunk res


genome_scatter <- function(df, title, padj_cutoff = 0.05) {
     df_sig <- df %>% filter(!is.na(padj), padj < padj_cutoff, !is.na(genome_order), !is.na(neg_log10_padj))
     target <- df_sig %>% filter(gene_id == "MSMEG_3117")
     
     ggplot(df_sig, aes(x = genome_order, y = log2FoldChange, colour = neg_log10_padj)) +
         geom_hline(yintercept = 0, linetype = "dashed", colour = "grey70") +
         geom_point(size = 1.6, alpha = 0.9) +
         geom_point(data = target, colour = "black", fill = "white", size = 3.5, shape = 21, stroke = 1) +
         geom_text_repel(
             data = target,
             aes(label = paste0(gene_id, "\nlog2FC = ", round(log2FoldChange, 2))),
             size = 3.3,
             nudge_y = 0.8,
             box.padding = 0.4,
             max.overlaps = Inf ) +
         scale_colour_gradient(low = "yellow", high = "red", name = expression(-log[10]~"adj p")) +
         labs(title = title, x = "Genome order (MSMEG number)", y = expression(log[2]~fold~change)) +
         theme_bw(base_size = 12) +
         theme(panel.grid.minor = element_blank())
}
p <- genome_scatter(
  df    = tidy_res,
  title = "Genome-wide DE (FDR < 0.05)"
)

print(p)

```

### Top DEG barplots

```{r}
plot_top_bar <- function(top_df, title) {
+     df <- top_df %>% mutate(gene = factor(gene, levels = rev(gene)))
+     
+     ggplot(df, aes(x = gene, y = log2FoldChange)) +
+         geom_col() +
+         coord_flip() +
+         labs(title = title, x = NULL, y = expression(log[2]~fold~change)) +
+         theme_classic(base_size = 12)}
```

### Heatmap

```{r}


sample_order <- c(
  "BR1_sgRNAneg_ATc",
  "BR2_sgRNAneg_ATc",
  "BR3_sgRNAneg_ATc",
  "BR1_sgRNApos_ATc",
  "BR2_sgRNApos_ATc",
  "BR3_sgRNApos_ATc"
)

stopifnot(all(sample_order %in% colnames(mat_rlog_rm)))

mat_ord <- mat_rlog_rm[, sample_order, drop = FALSE]
ann_ord <- coldata[sample_order, , drop = FALSE]


sample_cor <- cor(mat_ord, method = "pearson", use = "pairwise.complete.obs")


pheatmap::pheatmap(
  sample_cor,
  annotation_col = ann_ord,
  annotation_row = ann_ord,
  cluster_rows = FALSE,
  cluster_cols = FALSE,
  main = "Sample-to-sample correlation (rlog, batch removed)",
  fontsize = 10
)


rv <- matrixStats::rowVars(mat_ord)
rv <- rv[is.finite(rv)]

# remove zero-variance genes just in case
keep_genes <- names(rv)[rv > 0]
mat_use <- mat_ord[keep_genes, , drop = FALSE]
rv_use <- matrixStats::rowVars(mat_use)

top500 <- names(sort(rv_use, decreasing = TRUE))[1:500]
mat_top <- mat_use[top500, , drop = FALSE]

# z-score per gene (row)
mat_top_z <- t(scale(t(mat_top)))


pheatmap::pheatmap(
  mat_top_z,
  annotation_col = ann_ord,
  show_rownames = FALSE,
  cluster_cols = FALSE,   # keeps your specified sample order
  main = "Top 500 variable genes (rlog, batch removed)",
  fontsize = 10
)



```

```{r}
plot_heatmap_top500 <- function(dds_obj, main_title, filename) {
+     rld <- rlog(dds_obj, blind = FALSE)
+     rlog_mat <- assay(rld)
+     
+     rv <- apply(rlog_mat, 1, var, na.rm = TRUE)
+     top500 <- order(rv, decreasing = TRUE)[1:min(500, length(rv))]
+     mat_top500 <- rlog_mat[top500, , drop = FALSE]
+     
+     mat_top500_z <- t(scale(t(mat_top500)))
+     
+     ann_col <- as.data.frame(colData(dds_obj)[, c("condition", "replicate")])
+     
+     png(file.path(outdir, filename), width = 1400, height = 1100, res = 200)
+     pheatmap(mat_top500_z,
+              annotation_col = ann_col,
+              show_rownames = FALSE,
+              cluster_rows = TRUE,
+              cluster_cols = TRUE,
+              main = main_title)
+     dev.off() }
```

### Summary output

```{r}
cat("  Total significant genes   :", sum(res$padj < 0.05, na.rm=TRUE), "\n")
```

## MSMEG-MTB homology

As the Mycobacterial genomes are not fully annotated - M. smemgatis being the least annotated of all -, we can create a homology table with the orthologs of Mtb to better understand the genes within the dataset and link them to standardised functional categories.

```{r}
annot <- Mtb_Msm_ortholog_function_map

# make key consistent with gene_id

annot2 <- annot %>%
    mutate(MSMEG = ifelse(grepl("^MSMEG_", MSMEG),
                             MSMEG,
                             paste0("MSMEG_", MSMEG)))

res_c1_annot <- sgRNAposATc_vs_sgRNAnegATc_DESeq2 %>%
    left_join(annot2, by = c("gene_id" = "MSMEG"))

write.csv(
         res_c1_annot,
         file = file.path(outdir, "DESeq2_sgRNAposATc_vs_sgRNAnegATc_Mtb_annotated.csv"),
         row.names = FALSE
)
```

## Gene Enrichment

[GoSeq](https://bioconductor.org/packages/release/bioc/html/goseq.html), developed byÂ [Young et al. (2010)](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-2-r14), tests for the enrichment of Gene Ontology terms. This is GOseq **enrichment** GOseq corrects for **gene length bias** and can be used with non-GO annotations, i.e., Mtb functional category annotations.

```{r}


## Build gene_length from annotation table (Start/Stop)

# If annotation object is named differently, change mc2155_annotation
stopifnot(exists("mc2155_annotation"))

annot_cds <- mc2155_annotation[mc2155_annotation$Feature == "CDS", ]
annot_cds$gene_length <- abs(annot_cds$Stop - annot_cds$Start) + 1

gene_lengths <- annot_cds[, c("Locus", "gene_length")]
colnames(gene_lengths) <- c("MSMEG", "gene_length")

# Merge into res_annot_df if not already present
if (!("gene_length" %in% colnames(res_annot_df) && any(res_annot_df$gene_length > 0, na.rm=TRUE))) {
  res_annot_df <- merge(res_annot_df, gene_lengths, by = "MSMEG", all.x = TRUE)
}

sum(is.na(res_annot_df$gene_length))

FDR <- 0.05
cat_col <- "Mtb_Functional_Category"
stopifnot(cat_col %in% colnames(res_annot_df))
stopifnot("gene_length" %in% colnames(res_annot_df))

# DE vectors
de_up <- with(res_annot_df, !is.na(padj) & padj < FDR & log2FoldChange > 0)
de_dn <- with(res_annot_df, !is.na(padj) & padj < FDR & log2FoldChange < 0)
names(de_up) <- res_annot_df$MSMEG
names(de_dn) <- res_annot_df$MSMEG

# Bias vector
bias <- res_annot_df$gene_length
names(bias) <- res_annot_df$MSMEG

# gene2cat mapping (drop NA categories)
ok_cat <- !is.na(res_annot_df[[cat_col]]) & res_annot_df[[cat_col]] != ""
gene2cat <- split(res_annot_df[[cat_col]][ok_cat], res_annot_df$MSMEG[ok_cat])

# Universe: genes with both length and category
universe <- intersect(names(bias)[!is.na(bias)], names(gene2cat))
de_up_u <- de_up[universe]
de_dn_u <- de_dn[universe]
bias_u  <- bias[universe]
gene2cat_u <- gene2cat[universe]

# UP
pwf_up <- nullp(de_up_u, bias.data = bias_u)
gs_up <- goseq(pwf_up, gene2cat = gene2cat_u)
gs_up$padj <- p.adjust(gs_up$over_represented_pvalue, method = "BH")
gs_up <- gs_up[order(gs_up$padj), ]

# DOWN
pwf_dn <- nullp(de_dn_u, bias.data = bias_u)
gs_dn <- goseq(pwf_dn, gene2cat = gene2cat_u)
gs_dn$padj <- p.adjust(gs_dn$over_represented_pvalue, method = "BH")
gs_dn <- gs_dn[order(gs_dn$padj), ]

head(gs_up, 20)
head(gs_dn, 20)

write.csv(gs_up, file.path(out_goseq, paste0(prefix, "_GOseq_MtbFunctional_UP.csv")), row.names = FALSE)
write.csv(gs_dn, file.path(out_goseq, paste0(prefix, "_GOseq_MtbFunctional_DOWN.csv")), row.names = FALSE)

png(file.path(out_plots, paste0(prefix, "_GOseq_PWF_UP.png")), width = 900, height = 700, res = 150)
plotPWF(pwf_up)
dev.off()

png(file.path(out_plots, paste0(prefix, "_GOseq_PWF_DOWN.png")), width = 900, height = 700, res = 150)
plotPWF(pwf_dn)
dev.off()


```
